catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/util/impl/QueryResultPaginatorTest.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p/>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p/>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.util.impl;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Result;
+import ddf.catalog.data.impl.ResultImpl;
+import ddf.catalog.operation.impl.QueryImpl;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+import ddf.catalog.operation.impl.QueryResponseImpl;
+
+public class QueryResultPaginatorTest {
+
+    private CatalogFramework catalogFrameworkMock;
+
+    private QueryRequestImpl queryRequestMock;
+
+    private QueryResponseImpl queryResponseMock;
+
+    private QueryImpl queryMock;
+
+    private QueryResultPaginator paginator;
+
+    @Before
+    public void setUp() throws Exception {
+
+        catalogFrameworkMock = mock(CatalogFramework.class);
+        queryRequestMock = mock(QueryRequestImpl.class);
+        queryResponseMock = mock(QueryResponseImpl.class);
+        queryMock = mock(QueryImpl.class);
+        when(catalogFrameworkMock.query(queryRequestMock)).thenReturn(queryResponseMock);
+        when(catalogFrameworkMock.query(queryRequestMock)).thenReturn(queryResponseMock);
+        when(queryRequestMock.getQuery()).thenReturn(queryMock);
+    }
+
+    @Test
+    public void testhasNextIdempotency() throws Exception {
```

Comment:
✏️ Not exactly idempotent as it does force the underlying queue to be filled more on each call. You need a test to ensure that does not happen.